# Twirkshop

### a Twilio Workshop

## Me

Educator

You can find me around Internettown as craigsdennis

And...you're hiring 👋

## What's In It For You?

> "...teach a small group of technical people a technical skill of some kind."

> -- Clint

## What's In It For You?

> "This is where we get to know you as a teacher and witness your ability to break down technical concepts into something your fellow developers can understand...we'd like to see how you approach things."

> -- Clint

## I Love to Experiment

Shows

Improv Workshops

Fun is crucial

## Enter Analysis Paralysis

I can teach anything in any format for any amount of time

Whatever I pick I'm for sure gonna use a **Jupyter Notebook**.

Hey...wait a second...

## So Let's Get a Little Meta

Let's walk through different features of the Jupyter Project

Discuss how they work for various teaching styles. Like this one.

## Believe It or Not

You're looking at a notebook right now

Jupyter Notebooks can be exported. One format is [Reveal.js](https://revealjs.com) slideshows

## Reveal.js

A powerful open source JavaScript based presentation application

### You Can Dive Deeper

When you need to

Like one of those Choose Your Own Adventure books

### Tons of Features

Rich plugins and custom styling

Configurable hotkeys

Show help

Presenter Notes (_hotkey T_)

Discuss 🌮and 🌯

## Project Jupyter

It's not spelled wrong

**Ju**lia **Pyt**hon and **R**

Whatever

## Jupyter Notebook

Web based application

Narrate, execute, and display results

There are various **kernels** that handle the interactive code 

### Python Kernel

In [1]:
1 + 1

2

### Imagine You Haven't Seen This Before

This is making use of the powerful [Python List Comprehension](https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions).

In [2]:
ideas = ['scooters', 'cigarettes', 'bikes', 'mail', 'ink', 'trading', 'learning', 'sports',
         'eggs', 'gum', 'earning', 'skateboards', 'tattoos']
['e-' + idea for idea in ideas if not idea.startswith('e')]

['e-scooters',
 'e-cigarettes',
 'e-bikes',
 'e-mail',
 'e-ink',
 'e-trading',
 'e-learning',
 'e-sports',
 'e-gum',
 'e-skateboards',
 'e-tattoos']

## Shortcuts
* `Ctrl + Enter` executes a Cell
* `Alt + Enter` executes a Cell and adds a new blank below
* Enter `Command mode` with `esc`
* `a` adds a blank cell above and `b` adds a blank cell below
* `dd` deletes a Cell
* Use arrow keys and `j` and `k`
* Switch Cell types with `m` for markdown and `y` for code
* Enter `Edit mode` with Enter

## Cell Execution Order matters

Note the `In [X]` next to each Code Cell. Use this to see if things are in order as Cells don't execute automatically.

In [3]:
def hello(name="World"):
    print(f"Hello {name}")

Now if we edit the `hello` function to have a comma and an exclamation point, note how we need to re-execute the function to get the change

In [4]:
hello()

Hello World


## Environments

Every Jupyter Notebook has access to environment variables.

You can also install Python Packages to your environment, and they will be available.

In this example, I use the Python [Twilio API Client](https://pypi.org/project/twilio/).  I installed it into my environment using pip

In [5]:
from twilio.base.exceptions import TwilioException
from twilio.rest import Client

try:
    client = Client()
    phone_number = client.lookups.phone_numbers('+19713513060').fetch()
    print(f'The Country code is {phone_number.country_code} and formatted it is {phone_number.national_format}')
except TwilioException as err:
    print("You might need to set your environment variables. https://www.twilio.com/blog/2017/01/how-to-set-environment-variables.html")
    print(err)
          

The Country code is US and formatted it is (971) 351-3060


## Better Output

Instead of just printing, you can produce very rich content.

This ranges from graphs, audio, video and entire web applications.

In [6]:
from IPython.display import HTML
HTML('''
<style>
.twilio-app {
    border: 3px dashed #F22F46;
    color: #0D122B;
    padding: 30px;
}
</style>
''')

%%html
<div class="twilio-app">
    <h2>Example Twilio Sync Serverless Application</h2>
    <p>Text a good thing to (971) 351-3060</p>
    <ul id="synced">
    </ul>
</div>


## JavaScript

You can even tap into JavaScript.

Jupyter currently uses [require.js](https://requirejs.org/) for module management.

In [7]:
%%javascript
requirejs.config({
    paths: { 
        'twilioSync': ['//media.twiliocdn.com/sdk/js/sync/releases/0.8.5/twilio-sync.min'], 
    },                                         // strip .js ^, require adds it back
});

<IPython.core.display.Javascript object>

In [8]:
%%javascript
require(['twilioSync'], function(twilioSync) {
    function prependItems(items) {
        const el = document.getElementById('synced');
        items.forEach(item => {
            const li = document.createElement('LI');
            li.appendChild(document.createTextNode(item.data.value.body));
            el.prepend(li);
        });
    }
    async function syncMessages() {
        const response = await fetch('https://quartz-ant-5504.twil.io/sync-token');
        const json = await response.json();
        const sync = new twilioSync.SyncClient(json.token);
        const list = await sync.list('messages');
        const page = await list.getItems();
        prependItems(page.items);
        list.on('itemAdded', (evt) => prependItems([evt.item]));    
    };
    syncMessages();
});


<IPython.core.display.Javascript object>